test/retryOnEmpty.test.ts

```diff
@@ -0,0 +1,668 @@
+          stubBlockNumberRequest(),
+        ]);
+
+        await engine.handle(request);
+
+        expectProviderRequestNotToHaveBeenMade(sendAsyncSpy, request);
+      });
+    });
+  });
+});
+
+/**
+ * When using `stubProviderRequests` to list canned responses for specific
+ * requests that are made to `provider.sendAsync`, you don't need to provide the
+ * full request object to go along with the response, but only part of that
+ * request object. When `provider.sendAsync` is then called, we can look up the
+ * compare the real request object to the request object that was specified to
+ * find a match. This function is used to do that comparison (and other
+ * like comparisons).
+ *
+ * @param requestMatcher - A partial request object.
+ * @param request - A real request object.
+ * @returns True or false depending on whether the partial request object "fits
+ * inside" the real request object.
+ */
+function requestMatches(
+  requestMatcher: Partial<JsonRpcRequest<unknown>>,
+  request: JsonRpcRequest<unknown>,
+): boolean {
+  return (
+    Object.keys(requestMatcher) as unknown as (keyof typeof requestMatcher)[]
+  ).every((key) => isDeepStrictEqual(requestMatcher[key], request[key]));
+}
+
+/**
+ * Builds the JsonRpcEngine instance, the provider object, and the block
+ * tracker for use in tests.
+ */
+function createTestSetup() {
+  const engine = new JsonRpcEngine();
+  const provider = providerFromEngine(engine);
+  const blockTracker = new PollingBlockTracker({
+    provider: provider as Provider,
+  });
+
+  return { engine, provider, blockTracker };
+}
+
+/**
+ * Wraps code within a test that needs to make use of a JsonRpcEngine instance,
+ * a provider object, and/or a block tracker.
+ *
+ * @param args - Either an options bag and a function, or just a function. The
+ * options bag takes a single option, `withFinalMiddleware` — true by default —
+ * which places a final middleware in the middleware stack that ensures that the
+ * response is fully set and that the stack ends properly. The function is
+ * called with the JsonRpcEngine instance, provider, and block tracker.
+ * @returns Whatever the callback returns.
+ */
+async function withTestSetup<T>(
+  ...args:
+    | [WithTestSetupOptions, WithTestSetupCallback<T>]
+    | [WithTestSetupCallback<T>]
+) {
+  const [{ withFinalMiddleware = true }, fn] =
+    args.length === 2 ? [args[0], args[1]] : [{}, args[0]];
+  const setup = createTestSetup();
+
+  setup.engine.push(
+    createRetryOnEmptyMiddleware({
+      provider: setup.provider,
+      blockTracker: setup.blockTracker,
+    }),
+  );
+
+  if (withFinalMiddleware) {
+    setup.engine.push((req, res, _next, end) => {
+      if (res.id === undefined) {
+        res.id = req.id;
+      }
+
+      if (res.jsonrpc === undefined) {
+        res.jsonrpc = '2.0';
+      }
+
+      if (res.result === undefined) {
+        res.result = 'default response';
+      }
+
+      end();
+    });
+  }
+
+  return await fn(setup);
```

Comment:
For now there is no reason to have this function take a function — it could just return `setup`. However, this package relies on an older version of `eth-block-tracker`, so once we upgrade that package we can call `blockTracker.destroy` here, and that should make this follow the same sort of pattern that we've followed in other projects.